ospf6d/ospf6_auth_trailer.c

```diff
@@ -0,0 +1,820 @@
+	struct ospf6_neighbor *on = NULL;
+	struct ospf6_auth_hdr ospf6_auth_info;
+	uint16_t hdr_len = 0;
+	uint32_t oh_seqnum_h = 0;
+	uint32_t oh_seqnum_l = 0;
+
+	on = ospf6_neighbor_lookup(oh->router_id, oi);
+	hdr_len = ntohs(oh->length);
+	if (*pkt_len < hdr_len) {
+		if (IS_OSPF6_DEBUG_AUTH_RX)
+			zlog_warn("RECV[%s] Wrong %s packet auth length",
+				  oi->interface->name,
+				  ospf6_message_type(oh->type));
+		oi->at_data.rx_drop++;
+		return -1;
+	} else if (*pkt_len == hdr_len) {
+		/* no auth data in packet
+		 */
+		return -1;
+	}
+
+	switch (oh->type) {
+	case OSPF6_MESSAGE_TYPE_HELLO:
+		hello = (struct ospf6_hello *)((uint8_t *)oh
+					       + sizeof(struct ospf6_header));
+		if (OSPF6_OPT_ISSET_EXT(hello->options, OSPF6_OPT_AT)) {
+			if (on)
+				on->auth_present = true;
+		} else {
+			if (on)
+				on->auth_present = false;
+
+			if (oi->at_data.flags != 0) {
+				if (IS_OSPF6_DEBUG_AUTH_RX)
+					zlog_warn("RECV[%s] : Auth option miss-match in hello pkt",
+						  oi->interface->name);
+				oi->at_data.rx_drop++;
+			}
+
+			return -1;
+		}
+		break;
+	case OSPF6_MESSAGE_TYPE_DBDESC:
+		dbdesc = (struct ospf6_dbdesc *)((uint8_t *)oh
+						 + sizeof(struct ospf6_header));
+
+		if (OSPF6_OPT_ISSET_EXT(dbdesc->options, OSPF6_OPT_AT)) {
+			if (on)
+				on->auth_present = true;
+		} else {
+			if (on)
+				on->auth_present = false;
+
+			if (oi->at_data.flags != 0) {
+				if (IS_OSPF6_DEBUG_AUTH_RX)
+					zlog_warn("RECV[%s] : Auth option miss-match in DB desc pkt",
+						  oi->interface->name);
+				oi->at_data.rx_drop++;
+			}
+
+			return -1;
+		}
+		break;
+	case OSPF6_MESSAGE_TYPE_LSREQ:
+	case OSPF6_MESSAGE_TYPE_LSUPDATE:
+	case OSPF6_MESSAGE_TYPE_LSACK:
+		if ((on && on->auth_present == false)
+		    && (oi->at_data.flags != 0)) {
+			if (IS_OSPF6_DEBUG_AUTH_RX)
+				zlog_warn("RECV[%s] : Auth config miss-match in %s",
+					  oi->interface->name,
+					  ospf6_message_type(oh->type));
+			oi->at_data.rx_drop++;
+			return -1;
+		}
+		break;
+	default:
+		if (IS_OSPF6_DEBUG_AUTH_RX)
+			zlog_warn("RECV[%s] : Wrong packet type %d",
+				  oi->interface->name, oh->type);
+		return -1;
+	}
+
+	memset(&ospf6_auth_info, 0, sizeof(struct ospf6_auth_hdr));
+	memcpy(&ospf6_auth_info, (uint8_t *)oh + hdr_len, *pkt_len - hdr_len);
+	if (ntohs(ospf6_auth_info.length) > OSPF6_AUTH_HDR_FULL) {
+		if (IS_OSPF6_DEBUG_AUTH_RX)
+			zlog_warn("RECV[%s] : Auth config miss-match in %s",
+				  oi->interface->name,
+				  ospf6_message_type(oh->type));
+		oi->at_data.rx_drop++;
+		return -1;
+	}
+
+	/* after authentication header validation is done
+	 * reduce the auth hdr size from the packet length
+	 */
+	*at_len = ntohs(ospf6_auth_info.length);
+	*pkt_len = *pkt_len - *at_len;
+
+	if (on) {
+		oh_seqnum_h = ntohl(ospf6_auth_info.seqnum_h);
+		oh_seqnum_l = ntohl(ospf6_auth_info.seqnum_l);
+		if ((oh_seqnum_h >= on->seqnum_h) &&
+		     (oh_seqnum_l > on->seqnum_l)) {
+			/* valid sequence number received */
```

Comment:
This check is not correct.  Refer to middle of page 10 in RFC 7166:

> Upon reception, the sequence number MUST be greater than the **sequence number in the last accepted OSPFv3 packet of the same OSPFv3 packet type** from the sending OSPFv3 neighbor.  Otherwise, the OSPFv3 packet is considered a replayed packet and dropped. OSPFv3 packets of different types may arrive out of order if they are prioritized as recommended in [RFC4222].

The sequence number must be tracked separately for each different type of OSPFv3 message.

Reply:
The ch

> > When you restart ospf6d on router A. Neighborship on router B will go down and that results in reset of receive sequence number in router B.  This behaviour is same in some vendors who have implemented this feature.
> 
> It only works if router B "forgets" the sequence number for router A. **Router B is not supposed to forget the sequence number for router A.** Allowing the sequence number to be reset breaks replay protection. If you use another vendor for router B who implements this correctly, FRR `ospf6d` will not be able to establish an adjacency after a restart.


Can we take this up as a separate PR? We need to finalise on what method we will use to store the higher order sequence number in persistent memory, then implement the changes.
or If there is any existing method do suggest. I will do it.

Reply:
@eqvinox I just tested on the scenario that you mentioned. I see a strange behaviour where the sequence number stored in oi structure is not removed after u restart the ospf6d daemon. I had seen this while implementing the feature also. It's been sometime I implemented the feature so forgot to mention that. But I am not sure how ospf6_interface data is intact even after ospf6d restart. We can check with some old OSPF folks.